File: core/src/main/java/cloud/vinh/rebirthsaga/presentation/dungeon/WorldInputHandler.java
package cloud.vinh.rebirthsaga.presentation.dungeon;

import com.badlogic.gdx.Input;
import com.badlogic.gdx.InputAdapter;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.utils.viewport.Viewport;

/** World-level input, routed AFTER the UI stage by the screen's
 * {@code InputMultiplexer}. Touches that the stage consumed (any HUD control)
 * never reach this handler; only taps that fall through onto the world do, and
 * they are unprojected through the world viewport — including extend/letterbox
 * areas — before touching gameplay. World coordinates are y-up, so "up" is
 * +1 on the y axis. */
public final class WorldInputHandler extends InputAdapter {
    /** Callback that submits a move into the simulation for the active screen. */
    public interface MoveSink {
        void requestMove(int dx, int dy);
    }

    private final Viewport worldViewport;
    private final DungeonRenderer renderer;
    private final MoveSink sink;
    private final Vector3 unproject = new Vector3();

    public WorldInputHandler(Viewport worldViewport, DungeonRenderer renderer, MoveSink sink) {
        this.worldViewport = worldViewport;
        this.renderer = renderer;
        this.sink = sink;
    }

    @Override
    public boolean keyDown(int keycode) {
        switch(keycode) {
            case Input.Keys.UP:
            case Input.Keys.W:
                sink.requestMove(0, 1);
                return true;
            case Input.Keys.DOWN:
            case Input.Keys.S:
                sink.requestMove(0, -1);
                return true;
            case Input.Keys.LEFT:
            case Input.Keys.A:
                sink.requestMove(-1, 0);
                return true;
            case Input.Keys.RIGHT:
            case Input.Keys.D:
                sink.requestMove(1, 0);
                return true;
            default:
                return false;
        }
    }

    @Override
    public boolean touchDown(int screenX, int screenY, int pointer, int button) {
        unproject.set(screenX, screenY, 0f);
        worldViewport.unproject(unproject);
        int cellX = (int)Math.floor(unproject.x / DungeonRenderer.TILE_SIZE);
        int cellY = (int)Math.floor(unproject.y / DungeonRenderer.TILE_SIZE);
        // Targeting reads the last completed cell; while the move track is
        // animating, the screen's submit gate drops any command anyway.
        int dx = cellX - renderer.playerCellX();
        int dy = cellY - renderer.playerCellY();
        if(Math.abs(dx) + Math.abs(dy) == 1) {
            sink.requestMove(dx, dy);
            return true;
        }
        return false;
    }
}

File: core/src/main/kotlin/cloud/vinh/rebirthdungeon/presentation/dungeon/WorldInputHandler.kt
package cloud.vinh.rebirthdungeon.presentation.dungeon

import com.badlogic.gdx.Input
import com.badlogic.gdx.math.Vector3
import com.badlogic.gdx.utils.viewport.Viewport
import ktx.app.KtxInputAdapter
import kotlin.math.abs
import kotlin.math.floor

/** Stage gets first refusal. Taps and swipes resolve on release, so one gesture produces one command. */
class WorldInputHandler(private val viewport: Viewport, private val renderer: DungeonRenderer,
    private val wait: () -> Unit, private val sink: (Int, Int) -> Unit) : KtxInputAdapter {
    private var pointerId = -1
    private var startX = 0
    private var startY = 0
    fun cancelGesture() { pointerId = -1 }
    override fun touchCancelled(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
        if (pointer != pointerId) return false
        cancelGesture(); return true
    }
    override fun keyDown(keycode: Int): Boolean {
        when (keycode) {
            Input.Keys.UP, Input.Keys.W -> sink(0, 1)
            Input.Keys.DOWN, Input.Keys.S -> sink(0, -1)
            Input.Keys.LEFT, Input.Keys.A -> sink(-1, 0)
            Input.Keys.RIGHT, Input.Keys.D -> sink(1, 0)
            Input.Keys.SPACE, Input.Keys.PERIOD -> wait()
            else -> return false
        }
        return true
    }
    override fun touchDown(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
        if (pointerId != -1 || button != Input.Buttons.LEFT) return false
        pointerId = pointer; startX = screenX; startY = screenY
        return true
    }
    override fun touchUp(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
        if (pointer != pointerId) return false
        pointerId = -1
        val dx = screenX - startX; val dy = screenY - startY
        if (maxOf(abs(dx), abs(dy)) >= 32) {
            if (abs(dx) > abs(dy)) sink(if (dx > 0) 1 else -1, 0) else sink(0, if (dy < 0) 1 else -1)
        } else {
            val point = viewport.unproject(Vector3(screenX.toFloat(), screenY.toFloat(), 0f))
            val x = floor(point.x / DungeonRenderer.TILE_SIZE).toInt() - renderer.playerCellX
            val y = floor(point.y / DungeonRenderer.TILE_SIZE).toInt() - renderer.playerCellY
            if (abs(x) + abs(y) == 1) sink(x, y)
        }
        return true
    }
}
